### DISCUSSION TITLE
[Solved] Use DataTemplate for `ItemsControl.ItemsPanel`

### DISCUSSION THREAD

**Discussion by duck-dev** [AvaloniaUI/Avalonia#7268] [Q&A] [answered]
Hello everyone,

I am wondering if there is a way to use a DataTemplate for the `ItemsPanelTemplate` of the `ItemsControl.ItemsPanel`, to differentiate between two different templates, depending on the type of a specific property. So, if a specific property in the ViewModel for example is of TypeA, the template associated with the TypeA will be used and the same for type TypeB. 
So, actually the same as in https://docs.avaloniaui.net/docs/templates/data-templates. The problem however is, that I can't use a `ContentControl` element there, because it's not convertible to an `IPanel`, which the `ItemsPanelTemplate` of course expects. Any other type, that's in the inheritance tree of a Panel doesn't have a `Content` property, which I can use for the purpose of binding to the property. That's understandable, otherwise a Panel would inherit `ContentControl`.

Now, my question is: How would I go about achieving the same effect, but for an `ItemsPanelTemplate`.

Here, you can better imagine what I'm trying to achieve with an example. This would be almost valid. Only `ContentControl` is invalid in this context.
```xml
<ItemsControl Items="{Binding Items}">
    <ItemsControl.DataTemplates>
        <DataTemplate DataType="namespace:TypeA">
            <!-- Content A -->
        </DataTemplate>
        <DataTemplate DataType="namespace:TypeB">
            <!-- Content B -->
        </DataTemplate>
    </ItemsControl.DataTemplates>
    <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
            <ContentControl Content="{Binding SomeProperty}"/> <!-- This isn't possible, because it expects an `IPanel` -->
        </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
    
    <!-- ... -->
</ItemsControl>
```

**Comment by duck-dev** [comment]
After days of experimenting, I was still not able to find a way to get out of this rabbit hole. 
I had ideas to use resources, the `ContentPresenter` wrapped into a `Panel` parent and various other things, but all of them lead to a dead end in one way or another.

After trying to simply set the `Content` of the `ItemsPanelTemplate`, I've got this compiler error:
```
[XAMLIL] Unable to find the object initialization node inside deferred content, this shouldn't happen in default Xaml configuration, probably some AST transformer have broken the structure Line X, position Y.
```

This is the line it's referring to and I'm not quite sure why that's not valid, while the `CurrentPanel` property is of type `IPanel` and has a default value, which is a `WrapPanel` instance:
```xml
<ItemsPanelTemplate Content="{Binding CurrentPanel}"/>
```

And this is the property, but that should be alright. It's default value is set to a `WrapPanel` with a horizontal orientation.
```cs
protected internal IPanel? CurrentPanel
{
    get => _currentPanel;
    set => this.RaiseAndSetIfChanged(ref _currentPanel, value);
}
```

I would greatly appreciate a clarification and possibly a good recommendation on how to achieve my goal.

**Comment by robloo** [comment]
This concept usually requires a DataTemplateSelector implementation. Someone with more experience with Avalonia will need to comment how to do that (if it's possible).

https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.datatemplateselector?view=windowsdesktop-6.0

**Reply by maxkatz6** [reply]
Instead of ```DataTemplateSelector``` in Avalonia we have ```IDataTemplate``` interface which can be implemented as developer wants.
But in both frameworks it's not about panel template.

Though in Avalonia it's possible to implement custom ```ITemplate<IPanel>```, should work...

**Reply by maxkatz6** [reply]
Also from what I remember it wasn't possible to invalidate template depending on custom DataTemplateSelector property. Same as in Avalonia.

**Comment by maxkatz6** [comment]
@duck-dev for some reason it's hard to understand what exactly and why exactly you need it. Definitely not the usual task.

At first glance it might be something what I would create with two different styles which set different items panel template depending on different selectors. Or any other possible way to replace whole template.

But also you should be able to create custom ```ITemplate<IPanel>``` with avalonia property for binding. But I don't really think it will work, because you can't invalidate template from inside of the template.

**Comment by FoggyFinder** [comment]
It's unclear to me what are you trying to achieve. Different panels for different types? Tbh it sounds like you actually don't have to use `ItemsPanelTemplate` but create 2 totally separated DataTemplates with its own `ItemsControl`.

**Comment by duck-dev** [comment]
@maxkatz6 @FoggyFinder I'll try to explain it a bit more specific. I have buttons listed in this control. The user can choose between a "Grid"-view and a "List"-view. 
If the "Grid" design is selected, the buttons should be aligned in a grid pattern (exactly what the `WrapPanel` does) and their size is for example just 300 x 300. 
If the "List"-view is selected however, the buttons are supposed to be listed vertically, one element per row and the buttons stretch to the entire width (`HorizontalAlignment="Stretch"`), while their height is reduced a bit.
I managed to change the look of the buttons, that was not a problem at all. The only point where I'm not able to continue from is switching the panel from `WrapPanel` to `StackPanel` and vice-versa.

And no, I don't want to create two entire UserControls for that, because that would be way too much duplicated "code". The `ItemTemplate` stays pretty much the same.

Here are some pictures of the two expected behaviours:

**Grid-pattern (horizontal `WrapPanel`):**
(Don't mind the green button in the bottom right corner)
![grafik](https://user-images.githubusercontent.com/61479436/147814361-9eeebf78-6c5d-448f-a8dc-327bb9ee9cb8.png)

**List-pattern (vertical `StackPanel`):**
(Again, don't mind the green button in the bottom right corner)
![grafik](https://user-images.githubusercontent.com/61479436/147814448-56ff213e-357b-46a8-8ac3-76e775bd6192.png)

**Reply by maxkatz6** [reply]
I would recommend using styles to toggle between these two states. Or create some kind of attached property to replace items panel templates depending on its value. It will be more straightforward solution as for me.

**Reply by duck-dev** [reply]
@maxkatz6 Well, I would like to, but apparently it's not as easy as it seems. As I stated in my message 12 hours ago, it doesn't seem to be possible to use a Binding for the Content of the `ItemsPanelTemplate` there. Setting the `ItemsControl.ItemsPanel` with resources leads to a compiler error as well.

Or did I misunderstand what you mean with the attached property?

**Reply by FoggyFinder** [reply]
@maxkatz6 from my understanding even if we do change `ItemsPanelTemplate` it won't work since changing `ItemsPanel` property doesn't trigger anything. Otherwise it would be possible to just use converter, something like that

```xml
ItemsPanel="{Binding ViewType, Converter={StaticResource ResourceKey=converter}}"
```

```cs
    public class ViewTypePanelConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            var template = new ItemsPanelTemplate()
            {
                Content = new Func<IServiceProvider, ControlTemplateResult>( _ => {
                    var content =
                        value switch
                        {
                            ViewType.V => new StackPanel { Orientation = Orientation.Vertical },
                            ViewType.H => new StackPanel { Orientation = Orientation.Horizontal },
                            _ => throw new NotImplementedException(),
                        };
                   return new ControlTemplateResult(content, new NameScope());
                })
            };
            return template;
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
```

If I recall properly when I had to do something similar I've created separated templates for ``ItemsControl` like I said before. Maybe there is a better way but I wasn't able to find it.

**Reply by duck-dev** [reply]
I assume that's why my attempt to solve it with different styles didn't work, as described in the message below. I'm wondering why that is so. Shouldn't that trigger a change as everything else does too?

**Comment by duck-dev** [comment]
@maxkatz6 I assume you meant something like the following when you were referring to toggling it with styles. I tried it with some usual patterns, which usually work fine for me. In this case however, the selector **doesn't change**, even though the **`set`-accessor** of the property is **definitely being called** and the **value actually changes** from `false` to `true` or vice-versa. So, there is nothing that should prevent it from switching the selector, but it still **doesn't do** it. And other controls that rely on this property actually change the used selector, so I can be 100% sure that it changes.
The `IsViewGrid` property is being set in the constructor of the class and the `ItemsControl` actually acknowledges that and uses the correct selector, but when I change the value of the property again at a later point, the changes don't reflect on the used selector. So, it either uses the grid- or list-selector, depending which one has been set in the constructor through the value of the `IsViewGrid` property, but the problem is that it doesn't change later.

This is how I've built the system:
```xml
<ItemsControl Items={Binding Items} 
                       Classes.grid-panel="{Binding IsViewGrid}"
                       Classes.list-panel="{Binding !IsViewGrid}">
    <!-- Everything related to the ItemTemplate... -->
</ItemsControl>
```
with the styles being defined like that:
```xml
<Style Selector="ItemsControl.grid-panel">
    <Setter Property="ItemsPanel">
        <Setter.Value>
            <ItemsPanelTemplate>
                <WrapPanel Orientation="Horizontal"/>
            </ItemsPanelTemplate>
        </Setter.Value>
    </Setter>
</Style>
    
<Style Selector="ItemsControl.list-panel">
    <Setter Property="ItemsPanel">
        <Setter.Value>
            <ItemsPanelTemplate>
                <StackPanel Orientation="Vertical"/>
            </ItemsPanelTemplate>
        </Setter.Value>
    </Setter>
</Style>
```

And the `IsViewGrid` property, which is responsible for toggling the used selector looks like this:
```cs
protected bool IsViewGrid
{
    get => _isViewGrid;
    set => this.RaiseAndSetIfChanged(ref _isViewGrid, value);
}
```

Is it possible that an `ItemsControl` element just ignores a change of the bound property?

**Comment by ahopper** [comment, marked as answer]
The simplest way to do this is to just include the ItemsControl in the datatemplates i.e. have two itemscontrols, you can share the inner datatemplate for the items by declaring it further up the tree.

**Reply by duck-dev** [reply]
I can't exactly imagine how that would work, since neither the `ItemsControl`, nor the `ItemsControl.ItemsPanel` can be used with a `ContentControl` and I also have to define the ItemTemplate somewhere, but I'm not sure how I can share it between the two. The documentation doesn't really help me realize that.

**Reply by FoggyFinder** [reply]
@duck-dev  I've created a small sample to show one way to switch between templates:

[How to dynamically change panel of ItemsControl?](https://software.codidact.com/posts/285438)

Well, I personally don't quite like it but I wasn't able to come to anything better than that.

**Reply by duck-dev** [reply]
Thank you very much for your answer. This contribution actually helped me to understand what you mean and achieve something quite similar to this, which solved the problem.

I greatly appreciate all your efforts. I need to get used to working with DataTemplates again and get more creative when thinking about solutions with DataTemplates.
Thank you very much and I sincerely wish you all a happy and successful New Year!
